### PR TITLE
[BugFix] TensorDictPrimer updates spec instead of overwriting

### LINKED ISSUE #2331: [BUG] `TensorDictPrimer` overwrites nested specs in the environemnt
part of #2327

## The primer overwrites any nested spec

Consider an env with nested specs

```python
 env = VmasEnv(
        scenario="balance,
        num_envs=5,
    )
```

add to it a primer for a nested hidden state

```python
    env = TransformedEnv(
        env,
        TensorDictPrimer(
            {
                "agents": CompositeSpec(
                    {
                        "h": UnboundedContinuousTensorSpec(
                            shape=(*env.shape, env.n_agents, 2, 128)
                        )
                    },
                    shape=(*env.shape, env.n_agents),
                )
            }
        ),
    )
```

the primer code in https://github.com/pytorch/rl/blob/0063741839a3e5e1a527947945494d54f91bc629/torchrl/envs/transforms/transforms.py#L4649 will overwirite the observation spec instead of updating it, resulting in the loss of all the spec keys that previoulsy were in the "agents" spec

The same result is obtained with

```python
    env = TransformedEnv(
        env,
        TensorDictPrimer(
            {
               ("agents","h"): UnboundedContinuousTensorSpec(
                            shape=(*env.shape, env.n_agents, 2, 128)
                )
            }
        ),
    )
```

here, updating the spec instead of overwriting it should do the job

### PR DESCRIPTION
**Pull request by matteobettini**
fixes #2331

it also solves the last issue in #2327

### CODE REVIEW

File: torchrl/envs/transforms/transforms.py [hunk 1]
```diff
@@ -4638,15 +4638,20 @@ def transform_observation_spec(
             raise ValueError(
                 f"observation_spec was expected to be of type CompositeSpec. Got {type(observation_spec)} instead."
             )
-        for key, spec in self.primers.items():
-            if spec.shape[: len(observation_spec.shape)] != observation_spec.shape:
-                expanded_spec = self._expand_shape(spec)
-                spec = expanded_spec
+        try:
+            device = observation_spec.device
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:4642]: when is observation_spec.device returning an `RuntimeError`?
IIRC this should not happen anymore

**Review by vmoens — COMMENTED** [review]
Thanks for this, can you add a small test in `TestTensorDictPrimer`?
Just create a TDPrimer and an observation spec with nested specs, and call `transform_observation_spec` to check that the entries are updated accordingly.

**Comment by matteobettini** [comment]
I added the test

**Review by vmoens — APPROVED** [review]
LGTM

**Pull request merged**